lib/phoenix/token.ex

```diff
@@ -0,0 +1,120 @@
+      true
+
+  In that example we have a user's id, we generate a token and send
+  it to the client. When the client uses the token in a futher 
+  communication you verify the token, query the database and authorize 
+  the user. 
+
+  When using it with a socket a typical example might be:
+
+    defmodule MyChannel do
+      def join("my:" <> id = topic, %{token: token}, socket) do
+        case verify_token(socket, token) do
+        user_id ->
+          socket = assigns(socket, :user, Repo.get!(User, user_id))
+          reply socket, "join", %{}
+        :expired -> {:error, :auth_expired, socket}                                                                                                                                                           
+        :error -> {:error, :bad_auth, socket}
+        end
+      end
+    end
+
+  In this example the Phoenix.js client will be sending up the token
+  in the join command. 
+
+  If you want to provide an API to the user send it down to the user
+  by generating it in a controller or view action. 
+
+    def create(conn, params) do
+      user = User.create(params)
+      render conn, "user.json", %{token: gen_token(conn, user.id), user: user}
+    end
+
+    Then the client can use it all the way through. If you'd like to 
+    use the same token for API authorization then I suggest adding it
+    as a header to your authroized http requests
+    
+      Authorization: Bearer %{token}
+
+    Then create a plug to verify the token and authorize with the database.
+
+  """
+  alias Plug.Crypto.KeyGenerator
+  alias Plug.Crypto.MessageEncryptor
+
+  @doc """
+  Encrypts your data into a token you can send down to clients
+  """
+  def gen_token(context, data, opts \\ []) do
+    {secret, sign_secret, max_age, encoder} = get_endpoint(context) |> encryptor()
+
+    if Dict.has_key?(opts, :max_age) do
+      max_age = opts[:max_age]
+    end
+
+    if max_age do
+      exp = now_ms() + max_age
+    else
+      exp = nil
+    end
+
+    message = %{ 
+      data: data,
+      exp: exp
+    } |> encoder.encode!()
+    MessageEncryptor.encrypt_and_sign(message, secret, sign_secret)
+  end
+
+  @doc """
+  Decrypts the token into the originaly present data.
+  """
+  def verify_token(context, token) do
+    {secret, sign_secret, max_age, encoder} = get_endpoint(context) |> encryptor()
+    case MessageEncryptor.verify_and_decrypt(token, secret, sign_secret) do
+      :error -> :error
+      {:ok, message} -> 
+        %{ "data" => data, "exp" => exp} = Poison.decode!(message)
+        if exp < now_ms() do
+          :token_expired
+        else
+          data
+        end
+    end
+  end
+
+  defp get_endpoint(%Plug.Conn{} = conn), do: Phoenix.Controller.endpoint_module(conn)
+  defp get_endpoint(%Phoenix.Socket{} = socket), do: socket.endpoint
+
+  # Gathers configuration and generates the key secrets and signing secrets.
+  defp encryptor(endpoint) do
+    config = endpoint.config(:token_auth)
```

Comment:
Actually, since the salt must be explicit, I would get rid of this configuration. Just ask everything explicitly in `opts` and we ship with reasonable defaults: https://github.com/elixir-lang/plug/blob/master/lib/plug/session/cookie.ex#L60-L66